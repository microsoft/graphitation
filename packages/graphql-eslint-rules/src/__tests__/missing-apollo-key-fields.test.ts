/*
 * Taken from https://github.com/dotansimha/graphql-eslint/blob/300f73be802bdd06432a5df34939521d1ce0d93b/packages/plugin/tests/require-id-when-available.spec.ts
 * MIT license https://github.com/dotansimha/graphql-eslint/blob/300f73be802bdd06432a5df34939521d1ce0d93b/LICENSE
 */

import { TextDecoder } from "util";
global.TextDecoder = TextDecoder as any;

import {
  GraphQLRuleTester,
  ParserOptions,
} from "@graphql-eslint/eslint-plugin";
import missingApolloKeyFieldsRule, {
  REQUIRE_KEY_FIELDS_WHEN_AVAILABLE,
} from "../missing-apollo-key-fields";

const TEST_SCHEMA = /* GraphQL */ `
  type Query {
    hasId: HasId!
    noId: NoId!
    vehicles: [Vehicle!]!
    keyField: [KeyFieldType]!
    flying: [Flying!]!
  }
  type NoId {
    name: String!
  }
  interface Vehicle {
    id: ID!
  }
  type Car implements Vehicle {
    id: ID!
    mileage: Int
  }
  interface Flying {
    hasWings: Boolean!
  }
  type Bird implements Flying {
    id: ID!
    hasWings: Boolean!
  }
  type KeyFieldType {
    objectId: ID!
    name: String!
  }
  type HasId {
    id: ID!
    name: String!
  }
`;

const WITH_SCHEMA = {
  parserOptions: <ParserOptions>{
    schema: TEST_SCHEMA,
    operations: [
      `fragment HasIdFields on HasId {
        id
      }`,
    ],
  },
};
const ruleTester = new GraphQLRuleTester();

export const typePolicies = {
  KeyFieldType: {
    keyFields: ["objectId"],
  },
};

ruleTester.runGraphQLTests(
  "missing-apollo-key-fields",
  missingApolloKeyFieldsRule as any, // FIXME: Not casting this as any leads to ts(2590) error
  {
    valid: [
      {
        ...WITH_SCHEMA,
        code: `query { noId { name } }`,
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { hasId { id name } }`,
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { keyField { objectId name } }`,
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { hasId { ...HasIdFields } }`,
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { vehicles { id ...on Car { id mileage } } }`,
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { vehicles { ...on Car { id mileage } } }`,
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { flying { ...on Bird { id } } }`,
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { vehicles { id ...on Car { mileage } } }`,
        options: [{ typePolicies }],
      },
    ],
    invalid: [
      {
        ...WITH_SCHEMA,
        code: `query { hasId { name } }`,
        errors: [{ messageId: REQUIRE_KEY_FIELDS_WHEN_AVAILABLE }],
        options: [{ typePolicies }],
      },
      {
        ...WITH_SCHEMA,
        code: `query { keyField { id name } }`,
        errors: [{ messageId: REQUIRE_KEY_FIELDS_WHEN_AVAILABLE }],
        options: [{ typePolicies }],
      },
    ],
  }
);
