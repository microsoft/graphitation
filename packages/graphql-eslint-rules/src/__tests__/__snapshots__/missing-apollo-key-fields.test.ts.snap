// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"> 1 | query { hasId { name } }
    |               ^^^^^^^ Field(s) \\"id\\" must be selected (when available on a type). Please make sure to include it in your selection set!
If you are using fragments, make sure that all used fragments  specifies the field(s) \\"id\\"."
`;

exports[` 2`] = `
"> 1 | query { keyField { id name } }
    |                  ^^^^^^^^^^ Field(s) \\"objectId\\" must be selected (when available on a type). Please make sure to include it in your selection set!
If you are using fragments, make sure that all used fragments  specifies the field(s) \\"objectId\\"."
`;
